src/librustc/infer/const_variable.rs

```diff
@@ -0,0 +1,271 @@
+                None => c,
+            }
+        } else {
+            c
+        }
+    }
+
+    /// Creates a snapshot of the type variable state.  This snapshot
+    /// must later be committed (`commit()`) or rolled back
+    /// (`rollback_to()`).  Nested snapshots are permitted, but must
+    /// be processed in a stack-like fashion.
+    pub fn snapshot(&mut self) -> Snapshot<'tcx> {
+        Snapshot {
+            snapshot: self.values.start_snapshot(),
+            relation_snapshot: self.relations.snapshot(),
+        }
+    }
+
+    /// Undoes all changes since the snapshot was created. Any
+    /// snapshots created since that point must already have been
+    /// committed or rolled back.
+    pub fn rollback_to(&mut self, s: Snapshot<'tcx>) {
+        debug!("rollback_to{:?}", {
+            for action in self.values.actions_since_snapshot(&s.snapshot) {
+                if let sv::UndoLog::NewElem(index) = *action {
+                    debug!("inference variable _#{}t popped", index)
+                }
+            }
+        });
+
+        let Snapshot { snapshot, relation_snapshot } = s;
+        self.values.rollback_to(snapshot);
+        self.relations.rollback_to(relation_snapshot);
+    }
+
+    /// Commits all changes since the snapshot was created, making
+    /// them permanent (unless this snapshot was created within
+    /// another snapshot). Any snapshots created since that point
+    /// must already have been committed or rolled back.
+    pub fn commit(&mut self, s: Snapshot<'tcx>) {
+        let Snapshot { snapshot, relation_snapshot } = s;
+        self.values.commit(snapshot);
+        self.relations.commit(relation_snapshot);
+    }
+
+    /// Returns a map `{V1 -> V2}`, where the keys `{V1}` are
+    /// const-variables created during the snapshot, and the values
+    /// `{V2}` are the root variables that they were unified with,
+    /// along with their origin.
+    pub fn consts_created_since_snapshot(
+        &mut self,
+        s: &Snapshot<'tcx>
+    ) -> ConstVariableMap<'tcx> {
+        let actions_since_snapshot = self.values.actions_since_snapshot(&s.snapshot);
+
+        actions_since_snapshot
+            .iter()
+            .filter_map(|action| match action {
+                &sv::UndoLog::NewElem(index) => Some(ty::ConstVid {
+                    index: index as u32,
+                    phantom: PhantomData,
+                }),
+                _ => None,
+            })
+            .map(|vid| {
+                let origin = self.values.get(vid.index as usize).origin.clone();
+                (vid, origin)
+            })
+            .collect()
+    }
+}
+
+impl<'tcx> ut::UnifyKey for ty::ConstVid<'tcx> {
+    type Value = ConstVariableValue<'tcx>;
+    fn index(&self) -> u32 { self.index }
+    fn from_index(i: u32) -> Self { ty::ConstVid { index: i, phantom: PhantomData } }
+    fn tag() -> &'static str { "ConstVid" }
+}
+
+impl<'tcx> ut::UnifyValue for ConstVariableValue<'tcx> {
```

Comment:
Yeah, this entire `impl` is just universal boilerplate.

Reply:
Yep. As above, I think we ought to land as is, but we should definitely make some time to talk about how to do this better.